tools/fork-network/src/cli.rs

```diff
@@ -1,191 +1,723 @@
-                    let flat_head_for_this_shard = ready.flat_head.hash;
-                    if let Some(hash) = flat_head {
-                        if hash != flat_head_for_this_shard {
-                            panic!("Not all shards have the same flat head. Please reset the fork, and run the node for a little longer");
-                        }
-                    }
-                    flat_head = Some(flat_head_for_this_shard);
-                } else {
-                    panic!(
-                        "Flat storage is not ready for shard {}: {:?}. Please reset the fork, and run the node for longer",
-                        shard_uid, flat_storage_status
-                    );
-                }
-            }
-            println!("Forking from the flat storage final head: {}", flat_head.unwrap());
-            flat_head.unwrap()
-        };
+        let chain =
+            ChainStore::new(store.clone(), near_config.genesis.config.genesis_height, false);
 
-        let fork_head_block = store
-            .get_ser::<near_primitives::block::Block>(
-                DBCol::Block,
-                &borsh::to_vec(&fork_head).unwrap(),
-            )?
-            .unwrap();
+        let map = chain.get_all_block_hashes_by_height(fork_head.height)?;
+        let (_, hashes) = (*map).iter().next().unwrap();
+        let block_hash = *hashes.iter().next().unwrap();
+        let header = chain.get_block_header(&block_hash)?;
+        let epoch_id = header.epoch_id();
+        let num_shards = shard_layout.num_shards();
+        let block_height = header.height();
 
-        let prev_state_roots = fork_head_block
-            .chunks()
-            .iter()
-            .map(|chunk| chunk.prev_state_root())
-            .collect::<Vec<_>>();
-
-        // TODO(robin-near): Make this part customizable via some kind of config file.
-        let self_validator = near_config.validator_signer.as_ref().unwrap();
-        let self_account = AccountInfo {
-            account_id: self_validator.validator_id().clone(),
-            amount: 50_000 * NEAR_BASE,
-            public_key: self_validator.public_key(),
-        };
+        let state_roots: Vec<StateRoot> = (0..num_shards)
+            .map(|shard_id| {
+                let shard_uid = epoch_manager.shard_id_to_uid(shard_id, &epoch_id).unwrap();
+                let chunk_extra = chain.get_chunk_extra(&fork_head.hash, &shard_uid).unwrap();
+                *chunk_extra.state_root()
+            })
+            .collect();
 
-        let mut storage_mutator = StorageMutator::new(
-            epoch_manager.clone(),
-            &*runtime,
-            fork_head_block.header().epoch_id(),
-            *fork_head_block.header().prev_hash(),
+        tracing::info!(?epoch_id, ?block_height, block_height, ?state_roots);
+        let mut store_update = store.store_update();
+        store_update.set_ser(DBCol::Misc, b"EPOCH_ID", epoch_id)?;
```

Comment:
nit: since it's a global namespace within this misc column, what about prefixing these keys with something like `FORK_TOOL_`, since otherwise it looks like any other thing might be responsible for it.

Reply:
Done